`WebCore::colorFromCocoaColor` can sometimes fail to convert a color

rr-codes opened the pull request:
#### d4bad90961616ac6562ff5dd66f5c1f4906159d5
<pre>
`WebCore::colorFromCocoaColor` can sometimes fail to convert a color
<a href="https://bugs.webkit.org/show_bug.cgi?id=257405">https://bugs.webkit.org/show_bug.cgi?id=257405</a>
rdar://108643609

Reviewed by Aditya Keerthi.

The method `[UIColor getRed:green:blue:alpha:]` can fail if the given color is of an incompatible
color space, such as `NSColor Generic Gray Gamma 2.2 Profile colorspace 0.5 1`. Previously in the
ColorIOS implementation of `WebCore::colorFromCocoaColor`, we would just silently fail and return
an empty color.

This PR fixes this by checking if `[UIColor getRed:green:blue:alpha:]` is successful. If it is not,
we instead create a single colored pixel, and then extract the converted color from it to return.

* Source/WebCore/platform/ios/ColorIOS.mm:
(WebCore::colorFromCocoaColor):

Canonical link: <a href="https://commits.webkit.org/264619@main">https://commits.webkit.org/264619@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/29bcfd641c4031ccc94a082fcdf52f055dd271ca

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/8119 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/8407 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/8623 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/9778 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/8204 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/8125 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/10394 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/8318 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/11072 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/8264 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/9337 "Passed tests") | [⏳ 🧪 api-mac](https://ews-build.webkit.org/#/builders/API-Tests-macOS-EWS "Waiting to run tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/9900 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/6638 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/7427 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/14994 "5 flakes 136 failures") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/7761 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/7552 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/10910 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/8025 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/6524 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/7329 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1963 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/11537 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/7769 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/platform/ios/ColorIOS.mm:
```diff
@@ -28,6 +28,8 @@
 
 #if PLATFORM(IOS_FAMILY)
 
+#import "CGContextStateSaver.h"
```
Can remove this header.

pxlcoder commented on Source/WebCore/platform/ios/ColorIOS.mm:
```diff
@@ -45,7 +47,19 @@ Color colorFromCocoaColor(UIColor *color)
+        uint8_t pixel[4];
+        auto bitmapContext = adoptCF(CGBitmapContextCreate(pixel, 1, 1, 8, 4, sRGBColorSpaceRef(), kCGImageAlphaPremultipliedLast));
+
+        CGContextStateSaver stateSaver { bitmapContext.get() };
```
No need to save/restore state here.

pxlcoder commented on Source/WebCore/platform/ios/ColorIOS.mm:
```diff
@@ -45,7 +47,19 @@ Color colorFromCocoaColor(UIColor *color)
-    [color getRed:&redComponent green:&greenComponent blue:&blueComponent alpha:&alpha];
+    BOOL success = [color getRed:&redComponent green:&greenComponent blue:&blueComponent alpha:&alpha];
+    if (!success) {
+        // The color space conversion above can fail if the UIColor is in an unsupported colorspace.
```
`unsupported colorspace` -> `incompatible color space`

pxlcoder reviewed (commented):
> In Catalyst,

Remember that this is not just about Catalyst, the method can also return `NO` on iOS.

rr-codes committed:
`WebCore::colorFromCocoaColor` can sometimes fail to convert a color
https://bugs.webkit.org/show_bug.cgi?id=257405
rdar://108643609

Reviewed by Aditya Keerthi.

The method `[UIColor getRed:green:blue:alpha:]` can fail if the given color is of an incompatible
color space, such as `NSColor Generic Gray Gamma 2.2 Profile colorspace 0.5 1`. Previously in the
ColorIOS implementation of `WebCore::colorFromCocoaColor`, we would just silently fail and return
an empty color.

This PR fixes this by checking if `[UIColor getRed:green:blue:alpha:]` is successful. If it is not,
we instead create a single colored pixel, and then extract the converted color from it to return.

* Source/WebCore/platform/ios/ColorIOS.mm:
(WebCore::colorFromCocoaColor):

Canonical link: https://commits.webkit.org/264619@main
